int8 face
int32 angle
---
bool isFinish